README.md
# Hanami

Interactive arts and charts visualizations with Clojure(Script), Vega-lite, and Vega. Flower viewing 花見 (hanami)

<a href="https://jsa-aerial.github.io/aerial.hanami/index.html"><img src="https://github.com/jsa-aerial/hanami/blob/master/resources/public/Himeji_sakura.jpg" align="left" hspace="10" vspace="6" alt="hanami logo" width="150px"></a>

**Hanami** is a Clojure(Script) library and framework for creating interactive visualization applications based in [Vega-Lite](https://vega.github.io/vega-lite/) (VGL) and/or [Vega](https://vega.github.io/vega/) (VG) specifications. These specifications are declarative and completely specified by _data_ (JSON maps). VGL compiles into the lower level grammar of VG which in turn compiles to a runtime format utilizting lower level runtime environments such as [D3](https://d3js.org/), HTML5 Canvas, and [WebGL](https://github.com/vega/vega-webgl-renderer). In addition to VGL and VG, Hanami is built on top of [Reagent](http://reagent-project.github.io/) and [Re-Com](https://github.com/Day8/re-com).


Table of Contents
=================

   * [Overview](#overview)
   * [Installation](#installation)
   * [Features](#features)
   * [Examples](#examples)
      * [Simple cars](#simple-cars)
      * [Instrumented barchart](#instrumented-barchart)
      * [Contour plot using Vega template](#contour-plot-using-vega-template)
      * [Tree Layout using Vega template](#tree-layout-using-vega-template)
   * [Templates, Substitution Keys and Transformations](#templates-substitution-keys-and-transformations)
      * [Function values for substitution keys](#function-values-for-substitution-keys)
      * [Subtitution Key Functions](#subtitution-key-functions)
      * [Basic transformation rules](#basic-transformation-rules)
      * [Meta data and the :USERDATA key](#meta-data-and-the-userdata-key)
      * [Example predefined templates](#example-predefined-templates)
      * [Example predefined substitution keys](#example-predefined-substitution-keys)
      * [Data Sources](#data-sources)
         * [File data source](#file-data-source)
      * [Walk through example of transformation](#walk-through-example-of-transformation)
   * [Application Construction](#application-construction)
      * [Library overview](#library-overview)
      * [Framework overview](#framework-overview)
         * [framework topology graphic](#framework-topology-graphic)
      * [Resource requirements](#resource-requirements)
      * [Header](#header)
      * [Tabs](#tabs)
         * [Basics](#basics)
         * [Configuration and Behavior](#configuration-and-behavior)
         * [Extension tabs](#extension-tabs)
      * [Sessions](#sessions)
      * [Messages](#messages)
         * [Connection](#connection)
         * [Session group](#session-group)
         * [Tab updates](#tab-updates)
         * [User messages](#user-messages)
      * [Picture Frames](#picture-frames)
         * [Empty Frames](#empty-frames)
      * [Data Streaming](#data-streaming)
      * [Client Only Apps](#client-only-apps)
         * [Basic client app requirements](#basic-client-app-requirements)
            * [Require resources](#require-resources)
            * [Application initialization](#application-initialization)
            * [Landing page - index.html](#landing-page---indexhtml)
            * [Application start](#application-start)
         * [Example Client Only Application](#example-client-only-application)
      * [API](#api)
         * [Templates and Substitution keys](#templates-and-substitution-keys)
         * [Startup](#startup)
            * [Server start](#server-start)
            * [Client start](#client-start)
         * [Message system](#message-system)
            * [send msg](#send-msg)
            * [user msg](#user-msg)
         * [Client core](#client-core)
            * [Visualization](#visualization)
            * [Tab system](#tab-system)
            * [Hanami main](#hanami-main)
         * [Server core](#server-core)
   * [Example Transform 'Gallery'](#example-transform-gallery)
      * [Observed vs Binomial Model](#observed-vs-binomial-model)
      * [Lowess smoothing of Tn-Seq fitness data](#lowess-smoothing-of-tn-seq-fitness-data)
         * [With Overview   Detail](#with-overview--detail)
      * [Interactive Cross plot Differential Gene Expression](#interactive-cross-plot-differential-gene-expression)

[toc](https://github.com/ekalinin/github-markdown-toc)

# Overview

**Hanami** is a Clojure(Script) library and framework for creating interactive visualization applications based in [Vega-Lite](https://vega.github.io/vega-lite/) (VGL) and/or [Vega](https://vega.github.io/vega/) (VG) specifications. These specifications are declarative and completely specified by _data_ (JSON maps). VGL compiles into the lower level grammar of VG which in turn compiles to a runtime format utilizting lower level runtime environments such as [D3](https://d3js.org/), HTML5 Canvas, and [WebGL](https://github.com/vega/vega-webgl-renderer).

In keeping with the central data oriented tenet, Hanami eschews the typical API approach for generating specifications in favor of using recursive transforms of parameterized templates. This is also in keeping with the data transformation focus in functional programming, which is espcially nice in Clojure(Script).

An important aspect of this approach is that parameterized templates can be used to build other such templates by being higher level substitutions. In addition templates can be composed and this is an important idiomatic use. Furthermore, templates may be merged, though typically this is after transformation. The overall result enables the construction of sharable libraries of templates providing reusable plots, charts, and entire visualizations. Generally these will be domain and/or task specific. Hanami itself provides only a small set of very generic templates, which have proven useful in constructing more domain/task specific end results.

Hence, templates are a means to abstract all manner of visualization aspects and requirements. In this sense they are similar to what [Altair](https://altair-viz.github.io/) provides but without the complications and limitations of an OO class/method based approach.




# Installation

To install, add the following to your project `:dependencies`:

    [aerial.hanami "0.8.0"]




# Features

* Parameterized templates with recursive transformations
  * Takes the place of the typical static procedural/functional API
  * Purely data driven - no objects, classes, inheritance, whatever
  * Completely open ended - users may define their own with their own defaults
  * More general in scope than an API while capable of arbitrary specific detail
* A tabbing system for named visulization groupings
  * Multiple simultaneous independent and dependent visulizations per grouping
  * Automatic grid layout
  * Option system for customization
* Enables specific application construction
  * Application level page header instrumentation (re-com enabled)
  * Application level external instrumentation of charts (re-com enabled)
  * Full hiccup/re-com "picture framing" capability for independent charts
  * Multiple simultaneous (named) applications
  * Multiple sessions per application
    * Shared named sessions
  * Application extensible messaging capability
  * Data streaming capable
    * Realtime chart/plot updates with data updates
* Uses light weight websocket [messaging system](https://github.com/jsa-aerial/hanasu)




# Examples

```Clojure
(ns hanami.examples
  (:require [aerial.hanami.common :as hc]
            [aerial.hanami.templates :as ht]
            [aerial.hanami.core :as hmi]
            ...)
```

In all of the documentation, these namespaces are referred to by the shorthand provided in this require example.

* `aerial.hanami.common` == `hc`
* `aerial.hanami.templates` == `ht`
* `aerial.hanami.core` == `hmi`


## Simple cars

As a first example let's compare a Hanami template with corresponding Altair code. This is a the typical scatter plot example from the Vega-Lite developers used in many tutorials and by others in various places. First, the Altair:

```Python
cars = data.cars()

alt.Chart(cars).mark_point().encode(
    x='Horsepower',
    y='Miles_per_Gallon',
    color='Origin',
    ).interactive()
```

Hanami template version:

```Clojure
(hc/xform ht/point-chart
  :UDATA "data/cars.json"
  :X "Horsepower" :Y "Miles_per_Gallon" :COLOR "Origin")
```

Which, using the standard default [substitution keys](#example-predefined-substitution-keys) transforms to this Vega-Lite JSON specification:

```Clojure
{:data {:url "data/cars.json"},
 :width 400,
 :height 300,
 :background "floralwhite",
 :encoding
   {:x {:field "Horsepower", :type "quantitative"},
    :y {:field "Miles_per_Gallon", :type "quantitative"},
    :color {:field "Origin", :type "nominal"},
    :tooltip
    [{:field "Horsepower", :type "quantitative"}
     {:field "Miles_per_Gallon", :type "quantitative"}]}}
 ```


When rendered, both the Altair code and Hanami template, result in the following visualization, where the mouse is hovering over the point given by [132, 32.7]:

![Hanami pic 1](resources/public/images/hanami-cars-1.png?raw=true)


## Instrumented barchart

Hanami visualizations may be instrumented with external active componets (react/reagent/re-com) to enable external transforms on them. An example of an instrumented chart:

```Clojure
(hc/xform ht/bar-chart
  :USERDATA
  (merge
   (hc/get-default :USERDATA)
   {:vid :bc1
    :slider `[[gap :size "10px"] [label :label "Add Bar"]
              [label :label ~minstr]
              [slider
               :model :m1
               :min ~min, :max ~max, :step 1.0
               :width "200px"
               :on-change :oc1]
              [label :label ~maxstr]
              [input-text
               :model :m1
               :width "60px", :height "26px"
               :on-change :oc2]]})
  :HEIGHT 300, :WIDTH 350
  :X "a" :XTYPE "ordinal" :XTITLE "Foo" :Y "b" :YTITLE "Bar"
  :DATA data)
```

Which requires the active component implementing the instrument to be coded over on the client side. For this simple example, the test code on the client is a branch inside of a `cond` testing for the `:slider` key and value and then processing it:

```Clojure
      (let [cljspec spec
            udata (cljspec :usermeta)
            default-frame {:top [], :bottom [],
                           :left [[box :size "0px" :child ""]],
                           :right [[box :size "0px" :child ""]]}]
        (cond
          ...
          (udata :slider)
          (let [sval (rgt/atom "0.0")]
            (printchan :SLIDER-INSTRUMENTOR)
            (merge default-frame
                   {:top (xform-recom
                          (udata :slider)
                          :m1 sval
                          :oc1 #(do (bar-slider-fn tabid %)
                                    (reset! sval (str %)))
                          :oc2 #(do (bar-slider-fn tabid (js/parseFloat %))
                                    (reset! sval %)))}))
```

When this chart is rendered, it shows as (left, before slider move; right, after slider move)

![Hanami pic 2](resources/public/images/instrumented-chart-1a.png?raw=true)
![Hanami pic 3](resources/public/images/instrumented-chart-1b.png?raw=true)


## Contour plot using Vega template

An example using a Vega template for contour plotting. An important point to recognize here is that Vega specifications are also pure data, so the exact same recursive transformation works on Vega templates as Vega Lite templates:

```Clojure
(hc/xform
  ht/contour-plot
  :MODE "vega"
  :HEIGHT 400, :WIDTH 500
  :X "Horsepower", :XTITLE "Engine Horsepower"
  :Y "Miles_per_Gallon" :YTITLE "Miles/Gallon"
  :UDATA "data/cars.json"
  :XFORM-EXPR #(let [d1 (% :X)
                     d2 (% :Y)]
                 (format "datum['%s'] != null && datum['%s'] !=null" d1 d2)))
```

This generates far too much to show here, as Vega is a much lower level formal specification language than Vega Lite. This renders as:

![Hanami pic 3.1](resources/public/images/contour-1.png?raw=true)


## Tree Layout using Vega template

Another interesting Vega example uses the `tree-layout` template. Using this template for such layouts abstracts away a good deal of low level complexity. In this example a software system module dependency graph is rendered.

```Clojure
(hc/xform
 ht/tree-layout
  :MODE "vega"
  :WIDTH 650, :HEIGHT 1600
  :UDATA "data/flare.json"
  :LINKSHAPE "line" :LAYOUT "cluster"
  :CFIELD "depth")
 ```

![Hanami pic 3.2](resources/public/images/tree-layout-1.png?raw=true)




A number of other examples appear at the end of this README, along with their transformations and renderings.




# Templates, Substitution Keys and Transformations

_Templates_ are simply maps parameterized by _substitution keys_. Generally, templates will typically correspond to a legal VG or VGL specification or a legal subcomponent thereof. For example, a complete VGL specification (rendered as Clojure) is a legal template - even though it has no substitution keys. At the other extreme, templates can correspond to pieces of specifications or subcomponents. These will always have substitution keys - if they didn't there would be no point to them.

_Substitution Keys_ can be considered or thought of in two ways. They are the keys in the substitution map(s) of the recursive transformer [hc/xform](#templates-and-substitution-keys). There is a default map `hc/_defaults` (which is an atom containing the map) with various default substitution keys and their corresponding default values. This map can be updated and `xform` can also be supplied a variadic list of substitution keys and their values for a given invocation.

The second way of thinking about substitution keys is that they are the starting values of keys in templates. So, they represent parameterized values of keys in templates.

As an example consider the following. In a template we have a field and value `:field :X`. The `:X` here is a substitution key - it will be replaced as the value of `:field` during transformation by _its_ value in the current substitution map. In the default substitution map `hc/_defaults` `:X` has a value of "x", so the result will be `:field "x"`.

A more complex example would be the field and value `:encoding :ENCODING` in the predefined subcomponent `ht/view-base` which lays out the structure of a Vega-Lite _view_.  As before `:ENCODING` will be replaced during transformation by the value of `:ENCODING` in the substitution map. However, the default value of `:ENCODING` is the predefined subcomponent `ht/xy-encoding`. This value is a map describing the structure of a view's `x-y` encodings and contains many fields with their own substitution keys. So, to produce a final value, `ht/xy-encoding` is recursively transformed so that the final value of `:encoding` is a fully realized `x-y` encoding for the view being processed. `ht/view-base` has several other such fields and is itself recursively transformed in the context of the current substitution map. And its final value will be the base of a chart, such as a line chart (`ht/line-chart`) or area chart (`ht/area-chart`), or some new plot/chart/layout/etc of your own for some domain specific application.


## Function values for substitution keys

A substitution key may have a function of one argument for its value: `(fn[submap] ...)`. The `submap` parameter is the current substitution map in the transformation. This map contains the special key `::hc/spec` whose value is the initial (input) specification as well as all current substitution keys and their current values. This is useful in cases where a substitution key's value should be computed from other values. For example, a typical ([Saite](https://github.com/jsa-aerial/saite#user-tabs) does this by default) use case would be to compute the _label_ (display name) of a tab from its _id_:

```Clojure
        :TLBL #(-> :TID % name cljstr/capitalize)
```

This takes the tab's id, which here is a keyword, converts to a string and returns the capitalized version.


## Subtitution Key Functions

A more general version of the function as value of a substitution key is provided by the `hc/subkeyfns` map and associated processing. The function [hc/update-subkeyfns](#templates-and-substitution-keys) can be used to register a function for a subtitution key. This is a 'global' function - separate from any value given for the key. These functions take three arguments: `(fn[submap, subkey, subval] ...)`.

Where the `submap` parameter is the current substitution map in the transformation; the `subkey` parameter is the substitution key for this function; and the `subval` parameter is the _value_ of the substitution key.

These functions can be very useful in providing a further level of abstractions is template specifications. For example, it is very typical that a `color` specification is intended for the mark of a chart (line-chart, bar-chart, point-chart, etc). providing a simple string indicating the data element to facet with colors makes for a cleaner abstraction, but this needs to be converted to correct full form `field` and `type` element. Having a subkeyfn for the `hc/color-key` (default `:COLOR`) supports this sort of processing:

```Clojure
        color-key
         (fn[xkv subkey subval]
           (if (string? subval)
             (xform ht/default-mark-props (assoc xkv :MPFIELD subval))
             subval))
```

There is a default set of such keys provided with this particular case being one of them. There are a few others, including one which implements the [file](#file-data-source) data source capability for specifications.


## Basic transformation rules

There are some important rules that guide certain aspects of the recursive transformation process. These aspects are really only important in advanced template contruction, but if you find yourself in such circumstances, they are quite useful and helpful.

* Defaults for substitution keys are always overridden by values given for them in a call to [hc/xform](#templates-and-substitution-keys).

* Any field whose final value is the special `hc/RMV` value, will be removed from the corresponding specification or subcomponent thereof.

* Any field whose final value is an empty collection, will have this value replaced by `hc/RMV` and thus removed from its containing collection and specification. This is a very important transformation rule as it enables highly general substitution values. See for example such values as `ht/data-options`.

* While the simplest way to look at the transformation process is as a continuous transform until the last output is the same as the input, the actual processing is a preorder depth first walk and replacement. Typically, this implementation detail should not factor in to any template authoring, as they should be declarative in nature. However, there may be cases where this ordering can help in constructing proper substitution keys and values.


## Meta data and the `:USERDATA` key


CHANGELOG.md
# Change Log
All notable changes to this project, starting from 0.5.0, will be documented in this file.

## [0.8.1]
### Added
- Client: Support for user/app specific symbol translate on hiccup vecs
- Client: Visualization divs now use [:usermeta :vid] for id or gensym one.


## [0.8.0]
### Added
- Client: `hmi/sv!` as mirror of server side `hmi/sv!`. Client only apps support
- Client: `init-tabs` refactor `register`. Client only apps support
- Client: `frameit` refactor `vis-list`. Client only apps support
- Common: `:scaleFactor` to `:vgl` value in `hc/default-opts`. Support Vega-Embed export scaling
- Client: `make-frame` refactor instrumentor-fn handling.
- Client: `make-instrumentor` refactor making instrumentor - client only apps support
- Common: Add `:sort` support for default xy-encoding
- Common: Add tooltip title and format support for default-tooltip
- Common: Add generalized substitution keys for :XTTIP and :YTTIP
- Client: All frames now have an HTML id" (for getElementById)
- Client: Support for frame callbacks

### Fixed
- main rgt/render [hanami-main] did not use passed in element!

### Changed
- Client: internal name change user-dispatch to message-dispatch
- Server: internal name change server-dispatch to message-dispatch

### Deprecated


## [0.6.0]
### Added
- Client: dbg system
- Client: send-msg, replaces app-send
- Server: uuid DB
- Server: new clean send-msg (mirrors client)

### Changed
- Moved to Reagent 0.8.1

### Fixed
- Client: opts on registration was not xform'ed
- Client: remove bogus extra refs from re-com.box refer. Fixes shadow-cljs

### Deprecated
- Client: app-send, use send-msg


## [0.5.1]
### Added
- Template for over plus detail charts `ht/overview-detail. And new substitution key :DHEIGHT for controlling height of overview area. Experimental - may not work for all chart types
- :MODE "vega-lite" and :RENDERER "canvas" default substitution keys
- :XAXIS and :YAXIS substitution keys

### Changed
- x & y encoding :axis are now :XAXIS and :YAXIS. This is more general
- (default-opts :vgl) now uses :MODE and :RENDERER instead of hardcoded values

## [0.5.0]
### Added
- Chartless Picture frames (frames are hiccup and re-com enabled top, bottom, left, right areas surrounding a vis) Provides simple and easy support for full text based areas in tabs (chapters)
- New file data sources (previously only supported `values`, `url`, and `named` channels)
- New templates and default substitution keys to support tree and network layouts
- Stronger and more general support for adding application specific routes and landing page to Hanami websocket routes
- Nifty debug print system


[0.8.1]: https://github.com/your-name/hanami/compare/0.8.0...0.8.1
[0.8.0]: https://github.com/your-name/hanami/compare/0.6.0...0.8.0
[0.6.0]: https://github.com/your-name/hanami/compare/0.5.1...0.6.0
[0.5.1]: https://github.com/your-name/hanami/compare/0.5.0...0.5.1
[0.5.0]: https://github.com/your-name/hanami/compare/0.4.0...0.5.0


examples/basic_charts.clj
(ns hanami.basic-charts

  (:require [clojure.string :as str]
            #_[clojure.data.csv :as csv]
            [clojure.data.json :as json]

            #_[aerial.utils.math.probs-stats :as p]
            #_[aerial.utils.math.infoth :as it]

            #_[aerial.utils.string :as str]
            #_[aerial.utils.coll :as coll]

            [aerial.hanami.common :as hc :refer [RMV]]
            [aerial.hanami.templates :as ht]
            [aerial.hanami.core :as hmi]))



(defn log2 [x]
  (let [ln2 (Math/log 2)]
    (/ (Math/log x) ln2)))

(defn roundit [r & {:keys [places] :or {places 4}}]
  (let [n (Math/pow 10.0 places)]
    (-> r (* n) Math/round (/ n))))


(defn connfn [data]
  (assoc data :exps ["eid1" "eid2" "eid3"]))

(hmi/start-server
 3003
 :route-handler (hmi/hanami-handler
                 (hmi/hanami-routes :index-path "public/Fig/index.html"))
 :idfn (constantly "Exploring")
 :connfn connfn)
#_(hmi/stop-server)


(hc/update-defaults
 :USERDATA {:tab {:id :TID, :label :TLBL, :opts :TOPTS}
            :opts :OPTS
            :vid :VID, :msgop :MSGOP, :session-name :SESSION-NAME}
 :VID hc/RMV, :MSGOP :tabs, :SESSION-NAME "Exploring"
 :TID :expl1, :TLBL #(-> :TID % name str/capitalize)
 :OPTS (hc/default-opts :vgl), :TOPTS (hc/default-opts :tab))

(hc/xform (hc/get-default :USERDATA) :TID :geo :VID :v1)



;;; Simple scatter with template
(->> (hc/xform ht/point-chart
       ;;:DATA (->> "http://localhost:3003/data/cars.json" slurp json/read-str)
       :UDATA "data/cars.json"
       :X "Horsepower" :Y "Miles_per_Gallon" :COLOR "Origin")
     hmi/sv!)

;;; Same as above but 'template' == base vega-lite spec
(->>
 (hc/xform
  {:usermeta :USERDATA
   :data {:url "data/cars.json"},
   :mark "point",
   :encoding {:x {:field "Horsepower", :type "quantitative"},
              :y {:field "Miles_per_Gallon", :type "quantitative"},
              :color {:field "Origin", :type "nominal"}}})
 hmi/sv!)


;;; With picture framing
(let [text "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quod si ita est, sequitur id ipsum, quod te velle video, omnes semper beatos esse sapientes. Tamen a proposito, inquam, aberramus."
      top `[[gap :size "150px"]
            [p "An example showing a "
             [:span.bold "picture "] [:span.italic.bold "frame"]
             ". This is the top 'board'"
             [:br] ~text]]
      left `[[gap :size "10px"]
             [p {:style {:width "100px" :min-width "50px"}}
              "Some text on the " [:span.bold "left:"] [:br] ~text]]
      right `[[gap :size "2px"]
              [p {:style {:width "200px" :min-width "50px"
                          :font-size "20px" :color "red"}}
               "Some large text on the " [:span.bold "right:"] [:br]
               ~(.substring text 0 180)]]
      bottom `[[gap :size "200px"]
               [title :level :level3
                :label [p {:style {:font-size "large"}}
                        "Some text on the "
                        [:span.bold "bottom"] [:br]
                        "With a cool info button "
                        [info-button
                         :position :right-center
                         :info
                         [:p "Check out Saite Visualizer!" [:br]
                          "Built with Hanami!" [:br]
                          [hyperlink-href
                           :label "Saite "
                           :href  "https://github.com/jsa-aerial/saite"
                           :target "_blank"]]]]]]]
  (->> [(hc/xform ht/point-chart
          :TID :picframes
          :TOP top :BOTTOM bottom :LEFT left :RIGHT right
          :UDATA "data/cars.json"
          :X "Horsepower" :Y "Miles_per_Gallon" :COLOR "Origin")]
       hmi/sv!))

;;; With and without chart
(let [text "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quod si ita est, sequitur id ipsum, quod te velle video, omnes semper beatos esse sapientes. Tamen a proposito, inquam, aberramus."
      top `[[gap :size "50px"]
            [p "Here's a 'typical' chart/plot filled picture frame."
             "It only has the top area"
             [:br] ~text]]
      left `[[gap :size "20px"]
             [p {:style {:width "200px" :min-width "50px"}}
              "This is an empty frame with a " [:span.bold "left "]
              "column of text" [:br] ~text ~text ~text ~text]]
      right `[[gap :size "30px"]
              [p {:style {:width "200px" :min-width "50px"}}
               "And a " [:span.bold "right "]
               "column of text"
               [:br] ~text ~text ~text ~text]]]
  (->> [(hc/xform ht/point-chart
          :TID :picframes :UDATA "data/cars.json"
          :TOP top
          :X "Horsepower" :Y "Miles_per_Gallon" :COLOR "Origin")
        (hc/xform ht/empty-chart
          :TID :picframes
          :LEFT left :RIGHT right)]
       hmi/sv!))



;;; Simple Barchart with instrumented template
(->>
 (let [data [{:a "A", :b 28 },
             {:a "B", :b 55 },
             {:a "C", :b 43 },
             {:a "D", :b 91 },
             {:a "E", :b 81 },
             {:a "F", :b 53 },
             {:a "G", :b 19 },
             {:a "H", :b 87 },
             {:a "I", :b 52 }]
       min -10.0
       minstr (-> min str (str/split #"\.") first)
       max 10.0
       maxstr (-> max str (str/split #"\.") first (#(str "+" %)))]
   (hc/xform ht/bar-chart
             :USERDATA
             (merge
              (hc/get-default :USERDATA)
              {:vid :bc1
               :frame {:bottom `[p "Move slider to change data stream"]}
               :slider `[[gap :size "10px"] [label :label "Add Bar"]
                         [label :label ~minstr]
                         [slider
                          :model :m1
                          :min ~min, :max ~max, :step 1.0
                          :width "200px"
                          :on-change :oc1]
                         [label :label ~maxstr]
                         [input-text
                          :model :m1
                          :width "60px", :height "26px"
                          :on-change :oc2]]})
             :HEIGHT 300, :WIDTH 350
             :X "a" :XTYPE "ordinal" :XTITLE "Foo" :Y "b" :YTITLE "Bar"
             :DATA data))
 hmi/sv!)

(hmi/sd! {:usermeta {:msgop :data :vid :bc1}
          :data (mapv #(assoc % :b (+ 50 (% :b)))
                      [{:a "A", :b 28 },{:a "B", :b 55 },{:a "C", :b 43 },
                       {:a "D", :b 91 },{:a "E", :b 81 },{:a "F", :b 53 },
                       {:a "G", :b 19 },{:a "H", :b 87 },{:a "I", :b 52 }])})



(->>
 (hc/xform
  {:usermeta :USERDATA
   :height 600,
   :width 600,
   :data
   {:values
    [{:dose 0.5, :response 32659}
     {:dose 0.5, :response 40659}
     {:dose 0.5, :response 29000}
     {:dose 1, :response 31781}
     {:dose 1, :response 30781}
     {:dose 1, :response 35781}
     {:dose 2, :response 30054}
     {:dose 4, :response 29398}
     {:dose 5, :response 27779}
     {:dose 10, :response 27915}
     {:dose 15, :response 27410}
     {:dose 20, :response 25819}
     {:dose 50, :response 23999}
     {:dose 50, :response 25999}
     {:dose 50, :response 20999}]},
   :layer
   [{:selection {:grid {:type "interval", :bind "scales"}},
     :mark {:type "point", :filled true, :color "black"},
     :encoding
     {:x {:field "dose", :type "quantitative", :scale {:type "log"}},
      :y {:field "response", :type "quantitative", :aggregate "mean"}}}
    {:mark {:type "errorbar", :ticks true},
       :encoding
       {:x {:field "dose", :type "quantitative", :scale {:zero false}},
        :y {:field "response", :type "quantitative"},
        :color {:value "#4682b4"}}}]})
 hmi/sv!)

(->>
 (hc/xform
  {:usermeta :USERDATA
   :height 600,
   :width 600,
   :data
   {:values
    [{:Dose 0.5, :Response 32659.00003,
      :drc_dose 0.05, :drc_ll3 35597.08053881955},
     {:Dose 0.5, :Response 40659.00002340234,
      :drc_dose 1, :drc_ll3 35597.08053881955},
     {:Dose 0.5, :Response 29000,
      :drc_dose 2, :drc_ll3 35597.08053881955},
     {:Dose 1, :Response 31781,
      :drc_dose 5, :drc_ll3 35597.08053881955},
     {:Dose 1, :Response 30781,
      :drc_dose 10, :drc_ll3 35597.08053881955},
     {:Dose 1, :Response 35781,
      :drc_dose 50, :drc_ll3 35597.08053881955},
     {:Dose 2, :Response 30054,
      :drc_dose 200, :drc_ll3 35597.08053881955},
     {:Dose 4, :Response 29398,
      :drc_dose 1000, :drc_ll3 35597.08053881955}]},
   :layer
   [{:selection {:grid {:type "interval", :bind "scales"}},
     :mark {:type "point", :filled true, :color "black"},
     :encoding
     {:x {:field "Dose", :type "quantitative", :scale {:type "log"}},
      :y {:field "Response", :type "quantitative", :aggregate "mean"}}}
    {:mark {:type "errorbar", :ticks true},
       :encoding
       {:x {:field "Dose", :type "quantitative", :scale {:zero false}},
        :y {:field "Response", :type "quantitative"},
        :color {:value "black"}}}
    {:mark {:type "line", :color "red"},
     :encoding
     {:x {:field "drc_dose", :type "quantitative"},
      :y {:field "drc_ll3", :type "quantitative"}}}]})
 hmi/sv!)



;;; Geo Example
(->>
 (hc/xform
  {:usermeta :USERDATA
   :width 500,
   :height 300,
   :data {:url "data/airports.csv"},
   :projection {:type "albersUsa"},
   :mark "circle",
   :encoding {:longitude {:field "longitude", :type "quantitative"},
              :latitude {:field "latitude", :type "quantitative"},
              :tooltip [{:field "name", :type "nominal"}
                        {:field "longitude", :type "quantitative"}
                        {:field "latitude", :type "quantitative"}],
              :size {:value 10}},
   :config {:view {:stroke "transparent"}}}
  :TID :geo)
 hmi/sv!)



hc/subkeyfns
(hc/update-subkeyfns :TOOLTIP hc/RMV)

(->
 (let [data (->> (range 0.005 0.999 0.001)
                 (mapv (fn[p] {:x p, :y (- (log2 p)) :col "SI"})))]
   (hc/xform ht/line-chart
             :TID :multi :TOPTS {:order :row, :size "auto"}
             :TITLE "Self Information (unexpectedness)"
             :XTITLE "Probability of event", :YTITLE "-log(p)"
             :DATA data))
 hmi/sv!)




;;; Multi Chart - cols and rows
;;;
(->>
 [(let [data (->> (range 0.005 0.999 0.001)
                  (mapv (fn[p] {:x p, :y (- (log2 p)) :col "SI"})))]
    ;; Self Info - unexpectedness
    (hc/xform ht/layer-chart
              :TID :multi :TOPTS {:order :row, :size "auto"}
              :TITLE "Self Information (unexpectedness)"
              :LAYER [(hc/xform ht/line-layer
                                :XTITLE "Probability of event"
                                :YTITLE "-log(p)")
                      (hc/xform ht/xrule-layer :AGG "mean")]
              :DATA data))
  ;; Entropy - unpredictability
  (let [data (->> (range 0.00005 0.9999 0.001)
                  (mapv (fn[p] {:x p,
                               :y (- (- (* p (log2 p)))
                                     (* (- 1 p) (log2 (- 1 p))))})))]
    (hc/xform ht/layer-chart
              :USERDATA (merge (hc/get-default :USERDATA)
                               {:test2 [{:id 1 :label "One" :val 1}
                                        {:id 2 :label "Two" :val 2}
                                        {:id 3 :label "Three" :val 3}]})
              :TID :multi
              :TITLE "Entropy (Unpredictability)"
              :LAYER [(hc/xform ht/gen-encode-layer
                                :MARK "line"
                                :XTITLE "Probability of event" :YTITLE "H(p)")
                      (hc/xform ht/xrule-layer :AGG "mean")]
              :DATA data))]
 hmi/sv!)



;;; Some distributions
;;;;
(def obsdist
  (let [obs [[0 9] [1 78] [2 305] [3 752] [4 1150] [5 1166]
             [6 899] [7 460] [8 644] [9 533] [10 504]]
        totcnt (->> obs (mapv second) (apply +))
        pdist (map (fn[[k cnt]] [k (double (/ cnt totcnt))]) obs)]
    pdist))
;;(p/mean obsdist) => 5.7
(->>
 [(hc/xform ht/layer-chart
            :TID :dists :TOPTS {:order :row, :size "auto"}
            :TITLE "A Real (obvserved) distribution with incorrect simple mean"
            :HEIGHT 400 :WIDTH 450
            :LAYER
            [(hc/xform ht/bar-layer :XTITLE "Count" :YTITLE "Probability")
             (hc/xform ht/xrule-layer :AGG "mean")]
            :DATA (mapv (fn[[x y]] {:x x :y y :m 5.7}) obsdist))

  (hc/xform ht/layer-chart
            :TID :dists
            :TITLE "The same distribution with correct weighted mean"
            :HEIGHT 400 :WIDTH 450
            :LAYER
            [(hc/xform ht/bar-layer :XTITLE "Count" :YTITLE "Probability")
             (hc/xform ht/xrule-layer :X "m")]
            :DATA (mapv (fn[[x y]] {:x x :y y :m 5.7}) obsdist))]
 hmi/sv!)



;;; Contour maps (a Vega template!)
;;;
(->>
 (hc/xform
  ht/contour-plot
  :OPTS (merge (hc/default-opts :vgl) {:mode "vega"})
  :HEIGHT 400, :WIDTH 500
  :X "Horsepower", :XTITLE "Engine Horsepower"
  :Y "Miles_per_Gallon" :YTITLE "Miles/Gallon"
  :UDATA "data/cars.json"
  :XFORM-EXPR #(let [d1 (% :X)
                     d2 (% :Y)]
                 (format "datum['%s'] != null && datum['%s'] !=null" d1 d2)))
 hmi/sv!)

(->>
 (hc/xform
  ht/contour-plot
  :HEIGHT 500, :WIDTH 600
  :OPTS (merge (hc/default-opts :vgl) {:mode "vega"})
  :DATA (take 400 (repeatedly #(do {:x (rand-int 300) :y (rand-int 50)})))
  :XFORM-EXPR #(let [d1 (% :X)
                     d2 (% :Y)]
                 (format "datum['%s'] != null && datum['%s'] !=null" d1 d2)))
 hmi/sv!)




;;; =====================================================================;;;
